core/parachain/availability/fetch/fetch_impl.hpp

```diff
@@ -0,0 +1,51 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_PARACHAIN_AVAILABILITY_FETCH_FETCH_IMPL_HPP
+#define KAGOME_PARACHAIN_AVAILABILITY_FETCH_FETCH_IMPL_HPP
+
+#include "parachain/availability/fetch/fetch.hpp"
+
+#include <mutex>
+#include <unordered_map>
+
+#include "authority_discovery/query/query.hpp"
+#include "network/router.hpp"
+#include "parachain/availability/store/store.hpp"
+
+namespace kagome::parachain {
+  class FetchImpl : public Fetch,
+                    public std::enable_shared_from_this<FetchImpl> {
+   public:
+    FetchImpl(std::shared_ptr<AvailabilityStore> av_store,
+              std::shared_ptr<authority_discovery::Query> query_audi,
+              std::shared_ptr<network::Router> router);
+
+    void fetch(ValidatorIndex chunk_index,
+               const runtime::OccupiedCore &core,
+               const runtime::SessionInfo &session) override;
+
+   private:
+    struct Active {
+      ValidatorIndex chunk_index;
+      RelayHash relay_parent;
+      std::vector<primitives::AuthorityDiscoveryId> validators;
+      storage::trie::RootHash erasure_encoding_root;
+    };
+
+    void fetch(const CandidateHash &candidate_hash);
+    void fetch(const CandidateHash &candidate_hash,
```

Comment:
Trivial: avoid method overload whenever possible. Different reasonable name would be better.